src/NuGet.Services.AzureSearch/SearchService/StatusService.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Azure.Search.Models;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGet.Services.AzureSearch.Wrappers;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class StatusService : IStatusService
+    {
+        private readonly ISearchIndexClientWrapper _searchIndex;
+        private readonly ISearchIndexClientWrapper _hijackIndex;
+        private readonly IAuxiliaryDataCache _auxiliaryDataCache;
+        private readonly IOptionsSnapshot<SearchServiceConfiguration> _options;
+        private readonly ILogger<StatusService> _logger;
+
+        public StatusService(
+            ISearchIndexClientWrapper searchIndex,
+            ISearchIndexClientWrapper hijackIndex,
+            IAuxiliaryDataCache auxiliaryDataCache,
+            IOptionsSnapshot<SearchServiceConfiguration> options,
+            ILogger<StatusService> logger)
+        {
+            _searchIndex = searchIndex ?? throw new ArgumentNullException(nameof(searchIndex));
+            _hijackIndex = hijackIndex ?? throw new ArgumentNullException(nameof(hijackIndex));
+            _auxiliaryDataCache = auxiliaryDataCache ?? throw new ArgumentNullException(nameof(auxiliaryDataCache));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<StatusResponse> GetStatusAsync(Assembly assemblyForMetadata)
+        {
+            var response = new StatusResponse
+            {
+                Success = true,
+            };
+
+            await Task.WhenAll(
+                TryAsync(
+                    async () => response.SearchIndex = await GetIndexStatusAsync(_searchIndex),
+                    response,
+                    "warming the search index"),
+                TryAsync(
+                    async () => response.HijackIndex = await GetIndexStatusAsync(_hijackIndex),
+                    response,
+                    "warming the hijack index"),
+                TryAsync(
+                    async () => response.AuxiliaryFiles = await GetAuxiliaryFilesMetadataAsync(),
+                    response,
+                    "getting cached auxiliary data"),
+                TryAsync(
+                    async () => response.Server = await GetServerStatusAsync(assemblyForMetadata),
+                    response,
+                    "getting server information"));
+
+            return response;
+        }
+
+        private async Task TryAsync<T>(
+            Func<Task<T>> getAsync,
+            StatusResponse response,
+            string operation)
+        {
+            try
+            {
+                await Task.Yield();
```

Comment:
Why Task.Yield() is needed?

Reply:
So that all four of the status sub-operations run in parallel -- even leading, synchronous steps.